### PR TITLE
Connect DB and API

### PR DESCRIPTION
**Pull request by ann-kilzer**
Part of #87 

So there are some big changes here, let me know if you want to chat to square things up, especially regarding the schema changes.

# What changed
- [x] Connecting the SQL layer to the GraphQL
- [x] Remove mockData
- [x] Adjust some of the GraphQL models to match the DB data (I'll make comments inline so we can discuss)
- [x] Adjust Specialty DB to match GraphQL
- 🏉  Link healthcare professional to facility => another PR
- [x] healthcare professional (simplified version)

# Testing instructions

Sample queries for Apollo explorer are provided:

## Specialties

```graphql
query Specialties {
  specialties {
    id
    names {
      locale
      name
    }
  }
}
```

## Specialty by Id
```graphql
query Specialties($specialtyId: ID!) {
  specialty(id: $specialtyId) {
    id
    names {
      locale
      name
    }
  }
}

{
  "specialtyId": 1
}
```

## Facilities (missing HealthcareProfessional)

```graphql
query Facilities{
  facilities {
    contact {
      email
      mapsLink
      phone
      website
    }
    id
    nameEn
    nameJa
  }
}
```

## Facility by ID (missing HealthcareProfessional)

```graphql
query Facilities($facilityId: ID!){
  facility(id: $facilityId) {
    contact {
      email
      mapsLink
      phone
      website
    }
    id
    nameEn
    nameJa
  }

{
  "facilityId": 1
}
}
```

## HealthcareProfessional (simplified)

```graphql
query HealthcarePro{
  healthcareProfessionals {
    id
    names {
      firstName
      lastName
      locale
      middleName
    }
  }
}
```

## HealthcareProfessional by ID (simplified)

```graphql
($healthcareProfessionalId: ID!){
  healthcareProfessional(id: $healthcareProfessionalId) {
    names {
      firstName
      lastName
      locale
      middleName
    }
    id
  }
}

{
  "healthcareProfessionalId": 1
}
```

### CODE REVIEW

File: src/typeDefs/schema.graphql [hunk 1]
```diff
@@ -7,16 +7,12 @@ type Contact {
 
 type Facility {
   id: ID!
-  names: [FacilityName]!
```
Comment (ann-kilzer) [inline src/typeDefs/schema.graphql:10]: I simplified the facility names on the DB end and am adjusting the GraphQL to match.

My reasoning: Most medical websites in Japan have both a Japanese and an English or Romaji name. I haven't seen clinics using, say, Vietnamese names, etc., and I thought an overly complex model would be harder for MVP. Might be my own bias, and checking with Russ 🤔

File: src/typeDefs/schema.graphql [hunk 2]
```diff
@@ -49,22 +45,14 @@ type PersonName {
...
 
 input SpecialtyInput {
-  names: [SpecialtyNameInput]
+  nameEn: String
+  nameJa: String
 }
 
 type Specialty {
   id: ID!
-  names: [SpecialtyName]
```
Comment (ann-kilzer) [inline src/typeDefs/schema.graphql:57]: I should change this back... 🤔 we'll definitely want Specialties to have many translations 👀

File: prisma/migrations/20230114114951_specialties/migration.sql [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/*
```
Comment (ann-kilzer) [inline prisma/migrations/20230114114951_specialties/migration.sql:1]: I'm updating the database here to better match the GraphQL

File: prisma/schema.prisma [hunk 4]
```diff
@@ -12,18 +12,7 @@ datasource db {
   shadowDatabaseUrl = env("SHADOW_DB_URL")
 }
 
-// Represents the name of a human individual in both Japanese and Romanji characters.
```
Comment (ann-kilzer) [inline prisma/schema.prisma:15]: My original data model for names had an unnecessary level of indirection, and I simplified it to work better with the GraphQL

File: prisma/schema.prisma [hunk 5]
```diff
@@ -59,13 +47,23 @@ model SpokenLanguage {
 
 // Represents the area a professional works in, such as Gynecology, gastroenterology, oncology, physical therapy, etc.
 model Specialty {
-  id     Int    @id @default(autoincrement())
-  nameJa String
-  nameEn String
+  id    Int             @id @default(autoincrement())
```
Comment (ann-kilzer) [inline prisma/schema.prisma:50]: Model expanded to match the GraphQL

File: src/services/facilityService.ts [hunk 6]
```diff
@@ -1,13 +1,48 @@
...
+    contact: PrismaContact;
+})
+
+function convertPrismaToGraphQLFacility(input: FacilityAndRelations) {
+    return {
+        id: String(input.id),
+        nameJa: input.nameJa,
+        nameEn: input.nameEn,
+        contact: input.contact,
+        healthcareProfessionals: []
```
Comment (ann-kilzer) [inline src/services/facilityService.ts:16]: hydrating these objects is a bit complex. I'd love to work more on healthcareProfessionals in a following PR

File: prisma/migrations/20230114114951_specialties/migration.sql (hunk 3 above)
Comment (ann-kilzer) [inline prisma/migrations/20230114114951_specialties/migration.sql:1]: Reminder that migrations are generated files and will not be hand-edited. Please see `schema.prisma` for a clearer view of the changes

File: prisma/migrations/20230114134913_simplify_name/migration.sql [hunk 7]
```diff
@@ -0,0 +1,40 @@
+/*
```
Comment (ann-kilzer) [inline prisma/migrations/20230114134913_simplify_name/migration.sql:1]: Reminder that migrations are generated files and will not be hand-edited. Please see `schema.prisma` for a clearer view of the changes

File: src/typeDefs/schema.graphql [hunk 8]
```diff
@@ -27,21 +23,21 @@ input HealthcareProfessionalInput {
...
   id: ID!
-  names: [PersonName]!
+  names: [LocaleName]!
   degrees: [Degree]!
   spokenLanguages: [SpokenLanguage]!
   specialties: [Specialty]!
   acceptedInsuranceOptions: [Insurance]!
 }
 
-input PersonNameInput {
```
Comment (ann-kilzer) [inline src/typeDefs/schema.graphql:37]: Keeping in sync with the DB naming

File: prisma/schema.prisma (hunk 4 above)
Comment (ermish) [reply prisma/schema.prisma:15]: 💯  this makes more sense to me

File: prisma/schema.prisma [hunk 9]
```diff
@@ -59,13 +47,23 @@ model SpokenLanguage {
...
+  names SpecialtyName[]
 
   HealthcareProfessionalSpecialty HealthcareProfessionalSpecialty[]
 }
 
+model SpecialtyName {
```
Comment (ermish) [inline prisma/schema.prisma:56]: can we use `LocaleName` instead of SpecialtyName? Or does it need distinct fields that are different?

File: prisma/seed.ts [hunk 10]
```diff
@@ -45,13 +45,17 @@ async function seedSpecialties(verbose = false) {
             where: { id: index },
             update: {},
             create: {
-                nameEn: specialty[enCol],
-                nameJa: specialty[jaCol]
+                names: {
```
Comment (ermish) [inline prisma/seed.ts:48]: 💖

File: src/index.ts [hunk 11]
```diff
@@ -11,7 +11,6 @@ const server = new ApolloServer({
 const port = 3001;
 
 server.listen({ port }).then(({ url }) => {
+    // eslint-disable-next-line no-console
     console.log(`🚀  Server ready at ${url}`);
 });
-
-// add some lint
```
Comment (ermish) [inline src/index.ts:17]: I can't comment on the mockData file, but we should keep this. The main use case is for running automation tests.

File: src/services/facilityService.ts [hunk 12]
```diff
@@ -1,13 +1,48 @@
...
+    } as Facility;
+}
 
 // TODO: add a validation step for incoming parameters
-export const getFacilityById = (id: string) => {
-    const matchingFacility = facilities.find(
-        (location: Facility) => location.id === id
+export const getFacilityById = async (id: string) => {
+    const found = await prisma.facility.findUnique(
```
Comment (ermish) [inline src/services/facilityService.ts:22]: nit: could we call it something like `matchingFacility` or `facility`?

File: src/services/facilityService.ts [hunk 13]
```diff
@@ -1,13 +1,48 @@
...
+        {where: {
+            id: parseInt(id)
+        },
+        include: {
+            contact: true
+        }}
     );
 
-    return matchingFacility;
+    return found ? convertPrismaToGraphQLFacility(found) : null;
```
Comment (ermish) [inline src/services/facilityService.ts:31]: nit optimization: you could move the null check into the function so you don't have to keep doing the null check before calling

File: src/services/facilityService.ts [hunk 14]
```diff
@@ -1,13 +1,48 @@
...
 
-export const getFacilities = () => facilities;
+export const getFacilities = async () => {
+    const facilities = await prisma.facility.findMany(
+        {include: {
+            contact: true
+        }}
+    );
+
+    const ret = Array<Facility>();
```
Comment (ermish) [inline src/services/facilityService.ts:41]: nit: could we use more descriptive names? I couldn't figure out what `ret` was till I saw it was being returned. 

```suggestion
    const gqlFormattedFacilities = Array<Facility>();
```

File: src/services/facilityService.ts [hunk 15]
```diff
@@ -1,13 +1,48 @@
...
+
+    for (let i = 0; i < facilities.length; i++) {
```
Comment (ermish) [inline src/services/facilityService.ts:43]: personal preference nit: could use `facilities.foreach()` instead of `for()`

File: src/services/healthcareProfessionalService.ts [hunk 16]
```diff
@@ -1,13 +1,53 @@
...
+import { HealthcareProfessional, LocaleName } from '../typeDefs/gqlTypes';
+import { HealthcareProfessional as PrismaHealthcareProfessional,
+    LocaleName as PrismaLocaleName } from '@prisma/client';
+
+import prisma from '../db/client';
+
+type HealthcareProfessionalAndRelations = (PrismaHealthcareProfessional & 
+    { names: PrismaLocaleName[]})
+
+function convertPrismaToGqlHealthcareProfessional(input: HealthcareProfessionalAndRelations) {
```
Comment (ermish) [inline src/services/healthcareProfessionalService.ts:10]: idea for improving readability: not sure if we need this level of abstraction yet, but typically I put mappings into a separate folder to keep the services a bit cleaner.

File: src/services/healthcareProfessionalService.ts [hunk 17]
```diff
@@ -1,13 +1,53 @@
...
+        id: String(input.id),
+        names: Array<LocaleName>(),
+        specialties: [],
+        spokenLanguages: [],
+        acceptedInsuranceOptions: [],
+        degrees: []
+    } as HealthcareProfessional;
+
+    for (let i = 0; i < input.names.length; i++) {
+        ret.names.push(input.names[i] as LocaleName);
```
Comment (ermish) [inline src/services/healthcareProfessionalService.ts:22]:
```suggestion
        ret.names?.push(input.names[i] as LocaleName);
```

File: src/services/healthcareProfessionalService.ts [hunk 18]
```diff
@@ -1,13 +1,53 @@
...
+export const getHealthcareProfessionals = async () => {
+    const healthPros = await prisma.healthcareProfessional.findMany({
+        include: {
+            names: true
+        }
+    });
+
+    const ret = Array<HealthcareProfessional>();
+
+    for (let i = 0; i < healthPros.length; i++) {
```
Comment (ermish) [inline src/services/healthcareProfessionalService.ts:49]: actually, this is a perfect use case for `.map()`

```suggestion
    const gqlFormmatedEntities = healthPros.map(convertPrismaToGqlHealthcareProfessional)
```

File: src/services/specialtyService.ts [hunk 19]
```diff
@@ -1,13 +1,50 @@
...
+export const getSpecialtyById = async (id: string) => {
+    const found = await prisma.specialty.findUnique({ where: {
+        id: parseInt(id)
+    },
+    include: {
+        names: true
+    }});
 
-    return matchingSpecialty;
+    if (found) {
```
Comment (ermish) [inline src/services/specialtyService.ts:33]: readability recommendation: I recommend using the "[early return pattern](https://medium.com/swlh/return-early-pattern-3d18a41bba8)"

**Review by ermish — COMMENTED** [review]
Overall, looks awesome! It's coming together 🚀 
I saw a few changes that might be worth taking a look at before it's merged, but no major concerns

File: src/services/facilityService.ts (hunk 14 above)
Comment (ann-kilzer) [reply src/services/facilityService.ts:41]: Yeah this is just very `golang` style 😎 I can adjust

File: src/services/facilityService.ts (hunk 15 above)
Comment (ann-kilzer) [reply src/services/facilityService.ts:43]: I thought you asked us to be careful with foreach 🤔 

Is there a good rule of thumb for when to use each? I know foreach isn't always run in order

File: src/services/healthcareProfessionalService.ts (hunk 16 above)
Comment (ann-kilzer) [reply src/services/healthcareProfessionalService.ts:10]: It's a good idea, can we save for a future refactor?

File: src/services/facilityService.ts (hunk 12 above)
Comment (ann-kilzer) [reply src/services/facilityService.ts:22]: sure

File: src/services/facilityService.ts (hunk 13 above)
Comment (ann-kilzer) [reply src/services/facilityService.ts:31]: that's a good idea and now I'm wondering why i didn't think of it 🦞

File: src/services/facilityService.ts (hunk 14 above)
Comment (ann-kilzer) [reply src/services/facilityService.ts:41]: I'm going to try to find a happy medium between the golang `r` and `ret` and the Java `thisIsATemporaryVariableCounter` 

how about `gqlFacilites`

File: src/services/healthcareProfessionalService.ts (hunk 18 above)
Comment (ann-kilzer) [reply src/services/healthcareProfessionalService.ts:49]: Not sure if I follow, and some of the refactor suggestions are conflicting. I am gonna push some of the changes and let you comment again.

File: src/services/specialtyService.ts (hunk 19 above)
Comment (ann-kilzer) [reply src/services/specialtyService.ts:33]: I think that's a great pattern when there are nested ifs -- to break the logic out of nests.

But in this case, seems a bit "six of one, half-dozen of the other" 🤔

Comment (ann-kilzer) [reply src/services/specialtyService.ts:33]: Anyway it's about to disappear as part of another refactor

File: prisma/schema.prisma (hunk 9 above)
Comment (ermish) [reply prisma/schema.prisma:56]: oh the middle name and last is a good point. I'm sold 💰

File: src/index.ts (hunk 11 above)
Comment (ermish) [reply src/index.ts:17]: yep! ha, I think our first step is to start having some nice integration tests that actually use it!

File: src/services/facilityService.ts (hunk 14 above)
Comment (ermish) [reply src/services/facilityService.ts]: 💯 sounds great to me!

File: src/services/facilityService.ts (hunk 15 above)
Comment (ermish) [reply src/services/facilityService.ts]: sure! 

### use foreach when:
1) you want to do an operation on each record
2) order doesn't matter

### don't use foreach when:
1) you are changing the shape of the data or changing the order or size. You can use map, reduce, etc. 
2) the operation is async
3) you have 10,000+ records. It does have a minor performance hit on huge data sets

File: src/services/healthcareProfessionalService.ts (hunk 16 above)
Comment (ermish) [reply src/services/healthcareProfessionalService.ts]: yup! whenever you feel like it's starting to make the files long and hard to follow is probably the right time 😸

File: src/services/healthcareProfessionalService.ts (hunk 18 above)
Comment (ermish) [reply src/services/healthcareProfessionalService.ts]: ok! I'll add an example to the new changes

File: src/services/specialtyService.ts [hunk 20]
```diff
@@ -1,13 +1,50 @@
...
 
-export const getSpecialties = () => specialties;
+export const getSpecialties = async () => {
+    const dbSpecialties = await prisma.specialty.findMany({ include: {
+        names: true
+    }});
+
+    const gqlSpecialties = Array<Specialty>();
+
+    dbSpecialties.forEach(specialty => {
```
Comment (ermish) [inline src/services/specialtyService.ts:44]: Here's the example for how you could use `reduce()` here instead. Originally, you could use map to just change each record, but since it's conditionally adding, reduce makes more sense

```suggestion
    const gqlSpecialties  = dbSpecialties.reduce((convertedSpecialties, specialty) => {
        const gqlSpecialty = convertPrismaToGraphQLSpecialty(specialty);
         if (gqlSpecialty) { convertedSpecialties.push(gqlSpecialty); }         
         
         return convertedSpecialties;
}, []);
```

Comment (ermish) [reply src/services/specialtyService.ts:44]: ^ this is pretty much how most people do array conversions, filtering, etc nowadays. It's effectively the same functionality though, so this is totally up to you which way you prefer

**Review by ermish — APPROVED** [review]
LGTM! nothing blocking. take a look at the reduce example and see if it's something you'd want to update here

File: src/services/specialtyService.ts (hunk 20 above)
Comment (ann-kilzer) [reply src/services/specialtyService.ts:44]: I think it's cool and I'd like to study it more, but would feel better when I have better test coverage... right now nothing 😭 

How about I bookmark and return when there are unit or integration tests?
